SD_Card_Driver/Core/Src/LOS_D_sd.cpp

```diff
@@ -0,0 +1,96 @@
+#include "LOS_D_sd.hpp"
+
+int SDCard::init()
+{
+	return f_mount(&fs, "", 1);
+}
+
+int SDCard::open(const char* location)
+{
+	if (fileOpened && strcmp(location, filePath) == 0) {
+		return 0;
+	}
+
+	if (fileOpened) {
+		int res = f_close(&file);
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	int res = f_open(&file, location, FA_OPEN_ALWAYS | FA_WRITE | FA_READ);
+	fileOpened = (res == 0);
+
+	return res;
+}
+
+bool SDCard::checkExist(const char* location)
+{
+	int res = f_open(&file, location, FA_READ);
+	if (res == 0) {
+		f_close(&file);
+	}
+
+	return res == 0;
+}
+
+int SDCard::read(const char* location, uint8_t* data, size_t size, size_t offset)
+{
+	int res = this->open(location);
+	if (res != 0) {
+		return res;
+	}
+
+	res = f_lseek(&file, offset);
+	if (res != 0) {
+		return res;
+	}
+
+	size_t bytes_read;
+	res = f_read(&file, data, size, &bytes_read);
+
+	return res;
+}
+
+int SDCard::write(const char* location, uint8_t* data, size_t size, size_t offset)
+{
+	int res = this->open(location);
+	if (res != 0) {
+		return res;
+	}
+
+	res = f_lseek(&file, offset);
+	if (res != 0) {
+		return res;
+	}
+
+	size_t bytes_written;
+	res = f_write(&file, data, size, &bytes_written);
+
+	return res;
+}
+
+size_t SDCard::length(const char* location)
+{
+	int res = this->open(location);

```

Comment:
What the driver currently does is whenever read/write/length is called, it checks if the file at `location` is the file that is currently opened, if not then it closes the current file (if we currently have a file opened) and opens the file at `location`, then perform the read/write/length operation. The current file will remain open until the next operation closes it or the user calls `cleanup`.

The first method you suggested will be safer, but it's not optimal because we'll be opening and closing the file each time we call length/read/write.

The second method is probably better than what we currently have if we only allow one file to be open at a time, and the user remembers to close the file (I hope our EFS people are good programmers). If we want to support opening multiple files at once then it's more complicated. I'll explain in the next comment.

Reply:
I missunderstood how the system worked, I like this. In an ideal world I would probably advocate for a file object, where you instantiate the file object and that opens the file handle internally, then when the object is destroyed it gets closed.